### PR TITLE
Fix TPR watches.

### LINKED ISSUE #47027: Unable to watch for any ThirdPartyResource
**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.4.853+21b1fc9429aede-dirty", GitCommit:"21b1fc9429aedec76416a0e98c41d973b20acd67", GitTreeState:"dirty", BuildDate:"2017-06-05T04:49:24Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.4.852+7043372d05510b-dirty", GitCommit:"7043372d05510b13a0928af92deda6110740dd5e", GitTreeState:"dirty", BuildDate:"2017-06-01T22:26:46Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

Kubernetes v1.5.1 - v1.6.4 are **not** affected

**Environment**:
- **Cloud provider or hardware configuration**: baremetal/VM
- **OS** (e.g. from /etc/os-release): `Ubuntu 16.10`
- **Kernel** (e.g. `uname -a`): `4.9.17-040917-generic #201703220831`
- **Install tools**:
- **Others**:  
```
etcd --version
etcd Version: 3.1.0
Git SHA: 8ba2897
Go Version: go1.7.4
Go OS/Arch: linux/amd64
```

**What happened**:
It's impossible to perform a "watch" in any `ThirdPartyResource` created.

**What you expected to happen**:

To be able to "watch" for changes in any `ThirdPartyResource`

**How to reproduce it** (as minimally and precisely as possible):

1. Create a TPR as described here https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/#creating-a-thirdpartyresource

2. Run `kubectl get crontab -w` and leave the terminal open.

3. Create `my-crontab.yaml` as described in the above tutorial

4. Look at the terminal where you had `kubectl get crontab -w` open and realize `kubectl` exited without printing any error nor any object modification "watched".

Bonus:

Delete the object created in step 3., and run the following command in a new terminal:
```
curl 127.0.0.1:8080/apis/stable.example.com/v1/watch/crontabs
```

Run again step 3. and you can see `curl` as exited as well. You can't run the same `curl` command again unless you delete all objects from that TPR

### PR DESCRIPTION
**Pull request by enisoc**
Fixes #47027

TPR watch has been broken since #44350.

### CODE REVIEW

File: hack/make-rules/test-cmd-util.sh [hunk 1]
```diff
@@ -1629,15 +1629,32 @@ run_non_native_resource_tests() {
   # Make sure it's gone
   kube::test::get_object_assert foos "{{range.items}}{{$id_field}}:{{end}}" ''
 
+  # Test that we can watch the resource. Use coproc to read from stdout asynchronously.
+  coproc kubectl "${kube_flags[@]}" get bars --request-timeout=1m --watch-only -o name
```
Comment (deads2k) [inline hack/make-rules/test-cmd-util.sh]: @enisoc I haven't see this before.  Is it normal?

**Comment by deads2k** [comment]
question about normalcy of `coproc`, otherwise lgtm

File: hack/make-rules/test-cmd-util.sh (hunk 1 above)
Comment (enisoc) [inline hack/make-rules/test-cmd-util.sh]: It's a bash-only feature, but the hashbang at the top of the script specifies bash anyway:

https://www.gnu.org/software/bash/manual/html_node/Coprocesses.html

Comment (enisoc) [inline hack/make-rules/test-cmd-util.sh]: Do we have a shell style guide that says what bash version we can assume?

Comment (deads2k) [inline hack/make-rules/test-cmd-util.sh]: > Do we have a shell style guide that says what bash version we can assume?

mac os people get cranky, but we can let them sort it out I suppose.

**Comment by deads2k** [comment]
tests need to turn green.

/lgtm

**Pull request merged**